src/main/java/bio/terra/service/snapshot/SnapshotService.java

```diff
@@ -167,6 +169,40 @@ public SnapshotService(
     this.snapshotBuilderSettingsDao = snapshotBuilderSettingsDao;
   }
 
+  public String getSnapshotName(SnapshotRequestModel model) {
+    SnapshotRequestContentsModel contentsModel = model.getContents().get(0);
+    if (contentsModel.getMode() == SnapshotRequestContentsModel.ModeEnum.BYREQUESTID) {
+      SnapshotAccessRequestResponse snapshotAccessRequestResponse =
+          snapshotRequestDao.getById(contentsModel.getRequestIdSpec().getSnapshotRequestId());
+      String dashesAndSpacesRegex = "[- ]+";
+      String nonAlphaNumericRegex = "\\W";
+
+      // Handle null as empty string.
+      String snapshotAccessRequestName =
+          Optional.ofNullable(snapshotAccessRequestResponse.getSnapshotName()).orElse("");
+
+      String cleanedId = snapshotAccessRequestResponse.getId().toString().replace('-', '_');
+      String cleanedName =
+          StringUtils.truncate(
+              snapshotAccessRequestName
+                  .replaceAll(dashesAndSpacesRegex, "_")
+                  .replaceAll(nonAlphaNumericRegex, "")
+                  .trim(),
+              SNAPSHOT_NAME_MAX_LENGTH - 1 - cleanedId.length());
+      String separator = cleanedName.length() > 0 ? "_" : "";
+
+      String generatedName = String.format("%s%s%s", cleanedName, separator, cleanedId);
```

Comment:
This is fine although I would use `+` since there's no formatting going on
```suggestion
      String generatedName = cleanedName + separator + cleanedId;
```